test/e2e/lifecycle/recreate_node.go

```diff
@@ -0,0 +1,119 @@
+
+import (
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/labels"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/common"
+	"k8s.io/kubernetes/test/e2e/framework"
+	testutils "k8s.io/kubernetes/test/utils"
+)
+
+var _ = SIGDescribe("Recreate [Feature:Recreate]", func() {
+	f := framework.NewDefaultFramework("recreate")
+	var originalNodes []v1.Node
+	var originalPodNames []string
+	var ps *testutils.PodStore
+	systemNamespace := metav1.NamespaceSystem
+	BeforeEach(func() {
+		framework.SkipUnlessProviderIs("gce", "gke")
+		var err error
+		ps, err = testutils.NewPodStore(f.ClientSet, systemNamespace, labels.Everything(), fields.Everything())
+		numNodes, err := framework.NumberOfRegisteredNodes(f.ClientSet)
+		Expect(err).NotTo(HaveOccurred())
+		originalNodes, err = framework.CheckNodesReady(f.ClientSet, numNodes, framework.NodeReadyInitialTimeout)
+		Expect(err).NotTo(HaveOccurred())
+
+		framework.Logf("Got the following nodes before recreate %v", nodeNames(originalNodes))
+
+		ps, err = testutils.NewPodStore(f.ClientSet, systemNamespace, labels.Everything(), fields.Everything())
+		Expect(err).NotTo(HaveOccurred())
+		allPods := ps.List()
+
+		originalPods := filterIrrelevantPods(allPods)
+		originalPodNames = make([]string, len(originalPods))
+		for i, p := range originalPods {
+			originalPodNames[i] = p.ObjectMeta.Name
+		}
+
+		if !framework.CheckPodsRunningReadyOrSucceeded(f.ClientSet, systemNamespace, originalPodNames, framework.PodReadyBeforeTimeout) {
+			printStatusAndLogsForNotReadyPods(f.ClientSet, systemNamespace, originalPodNames, originalPods)
+			framework.Failf("At least one pod wasn't running and ready or succeeded at test start.")
+		}
+
+	})
+
+	AfterEach(func() {
+		if CurrentGinkgoTestDescription().Failed {
+			// Make that addon/system pods are running, so dump
+			// events for the kube-system namespace on failures
+			By(fmt.Sprintf("Collecting events from namespace %q.", systemNamespace))
+			events, err := f.ClientSet.CoreV1().Events(systemNamespace).List(metav1.ListOptions{})
+			Expect(err).NotTo(HaveOccurred())
+
+			for _, e := range events.Items {
+				framework.Logf("event for %v: %v %v: %v", e.InvolvedObject.Name, e.Source, e.Reason, e.Message)
+			}
+		}
+		if ps != nil {
+			ps.Stop()
+		}
+	})
+
+	It("recreate nodes and ensure they function upon restart", func() {
+		testRecreate(f.ClientSet, ps, systemNamespace, originalNodes, originalPodNames)
+	})
+})
+
+// Recreate all the nodes in the test instance group
+func testRecreate(c clientset.Interface, ps *testutils.PodStore, systemNamespace string, nodes []v1.Node, podNames []string) {
+	numNodes := len(nodes)
+	err := common.RecreateNodes(c, nodes)
+
+	if err != nil {
+		framework.Failf("Test failed; failed to recreate at least one node in the time given.")
```

Comment:
The split of responsibility seems a bit odd to me; so `RecreateNodes` is responsible for restarting the nodes and waiting for _one_ to restart? Looking at the exact implementation I can see you're waiting for the bootID to change, but to a caller it seems like they will either just want to fire/forget and not wait at all or restart and wait for the full restart to finish. I'm not a consumer of this though; so what are your thoughts on that?

> in the time given

Not clear how much time that is or where it comes from just by looking at the error message or the function call above it. Seems like we should pass the timeout into that method; if this recreation flow gets extended to other providers they may have different restart characteristics and want to fail faster/slower, especially in tests.


It seems like the more clear implementation would involve a method to recreate the nodes (and no waiting) and then maybe another method that will recreate and wait for them all to come back up with explicit timeouts passed to the method.

Reply:
Thanks for the suggestion. I agree that separating the restart and waiting is clearer for users and have changed that. 

However, it looks to me like the part of the function waiting for the boot ids to change waits for all the nodes to restart, as opposed to a single one. Could you clarify where it is waiting for only one node to restart?

Reply:
No you're right; so it was confusing since:
 - the function would _fail_ at the first one that fails
 - the error message was later interpreted as `failed to recreate at least one node`

So in the happy path, it did wait to see that each was updated.

Reply:
I see what you mean now. I think with waiting for all the node tries, this may be clearer?